// @hmcts/properties-volume/lib/Properties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const nodejs_logging_1 = require("@hmcts/nodejs-logging");
const path = require("path");
const fs = require("fs");
const { merge } = require('lodash');
const log = nodejs_logging_1.Logger.getLogger('applicationRunner');
const defaultOptions = {
    mountPoint: '/mnt/secrets/',
    failOnError: false
};
function addTo(config, givenOptions) {
    const options = merge({}, defaultOptions, givenOptions || {});
    const mountPoint = options.mountPoint;
    const failOnError = options.failOnError;
    log.info(`Attempting to read properties from volume: '${mountPoint}'`);
    try {
        const prefix = getPrefix(mountPoint.toString());
        const properties = readVaults(mountPoint);
        config[prefix] = merge(config[prefix] || {}, properties);
    }
    catch (error) {
        if (failOnError) {
            throw Error(`properties-volume failed with:'${error}`);
        }
        log.info(`Could not read properties from volume: '${mountPoint}' due to '${error}'`);
    }
    return config;
}
exports.addTo = addTo;
function getPrefix(aPath) {
    const prefixFolder = path.basename(aPath);
    if (prefixFolder.length === 0) {
        throw new Error(`Invalid properties mount point supplied: '${aPath}'`);
    }
    return prefixFolder;
}
function addDir(dir, obj, mountPoint) {
    obj[dir] = readDirectories(mountPoint, dir).reduce((values, file) => addFile(values, file, mountPoint, dir), {});
    return obj;
}
function addFile(values, file, mountPoint, dir) {
    values[file] = readFile(mountPoint, dir, file).trim();
    return values;
}
function readFile(mountPoint, dir, file) {
    return fs.readFileSync(mountPoint + '/' + dir + '/' + file, 'utf8');
}
function readDirectories(mountPoint, dir) {
    return fs.readdirSync(mountPoint + '/' + dir);
}
function readVaults(mountPoint) {
    return fs.readdirSync(mountPoint, null).reduce((obj, dir) => addDir(dir, obj, mountPoint), {});
}
